### PR TITLE
Various additions to make pure X work on Managarm

### PR DESCRIPTION
**Pull request by Dennisbonke**
This PR adds mlibc work that eventually should allow X.org to run in pure X mode on Managarm. As we're kinda stuck at the moment, this should be merged now, and any additional work can go into it's own PR. Notably lacking is a regex suite, we should settle on one to copy or implement our own.

### CODE REVIEW

**Comment by ArsenArsen** [comment]
dreadful, I love it

**Comment by avdgrinten** [comment]
I suggest to split the regex stuff off from this PR. Also, the regex engine looks like it is taken from elsewhere, w/o proper attribution (which is a copyright violation, even if its own on your private fork).

**Comment by Dennisbonke** [comment]
> I suggest to split the regex stuff off from this PR. Also, the regex engine looks like it is taken from elsewhere, w/o proper attribution (which is a copyright violation, even if its own on your own fork).

Yeah I didn't do this the smartest way. The regex engine here likely shouldn't be merged at all and indeed I messed up the attribution. I will fix this in the next few hours as I'm currently on the road. I'll eventually fix this PR so it conforms more to standards on other points, at which point the regex stuff will indeed be split off or dropped all together in favor of a better solution.

**Comment by Dennisbonke** [comment]
Addressed copyright violation, thanks again Alexander for bringing it to my attention.

**Comment by no92** [comment]
Merging 107992d as-is would break weston versions starting from mid-2019 forwards. Instead of stubbing, a complete implementation including the managarm side would be necessary.

File: options/posix/generic/sys-resource-stubs.cpp [hunk 1]
```diff
@@ -3,15 +3,17 @@
...
 	__ensure(!"Not implemented");
 	__builtin_unreachable();
 }
+
 int setpriority(int, id_t, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+	mlibc::infoLogger() << "mlibc: setpriority is a stub!" << frg::endlog;
+	return 0;
 }
```
Comment (avdgrinten) [inline options/posix/generic/sys-resource-stubs.cpp:17]: It'd be better to stub this on the sysdep level. Or does Linux implement this on top of a different syscall?

File: options/posix/generic/unistd-stubs.cpp [hunk 2]
```diff
@@ -16,8 +16,8 @@
 #include <mlibc/thread.hpp>
 
 unsigned int alarm(unsigned int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+	mlibc::infoLogger() << "mlibc: alarm is a stub!" << frg::endlog;
+	return 0;
 }
```
Comment (avdgrinten) [inline options/posix/generic/unistd-stubs.cpp:21]: This should be implemented on top of `setitimer`.

File: options/posix/generic/sys-resource-stubs.cpp (hunk 1 above)
Comment (Dennisbonke) [reply options/posix/generic/sys-resource-stubs.cpp]: Stubbed with a `sys_setpriority` syscall.

**Review by no92 — APPROVED** [review]

File: options/posix/generic/unistd-stubs.cpp [hunk 3]
```diff
@@ -15,9 +15,10 @@
...
 #include <mlibc/thread.hpp>
 
-unsigned int alarm(unsigned int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+unsigned int alarm(unsigned int seconds) {
+	struct itimerval it = {it.it_value.tv_sec = seconds}, old = {0};
+	setitimer(ITIMER_REAL, &it, &old);
+	return old.it_value.tv_sec + !! old.it_value.tv_usec;
 }
```
Comment (avdgrinten) [inline options/posix/generic/unistd-stubs.cpp:22]: What if `setitimer` fails here?

Comment (Dennisbonke) [reply options/posix/generic/unistd-stubs.cpp:22]: Musl doesn’t seem to handle that explicitly, and I guess setitimer takes care of errno in that case. On that note, the man page of alarm doesn’t list what should happen with respect to errors, nor does it list valid errno’s.

Comment (Dennisbonke) [reply options/posix/generic/unistd-stubs.cpp:22]: Hmm I’m not so sure about that. We could check the return value of setitimer and return 0 if it failed, but flat out crashing seems not the right solution.

Comment (Geertiebear) [reply options/posix/generic/unistd-stubs.cpp:22]: There are only two failing conditions for ``setitimer()``, which is if the pointers aren't valid or ``which`` is not one of ``ITIMER_REAL``, ``ITIMER_VIRTUAL``, or ``ITIMER_PROF``, neither of which happens in this case.

**Review by Geertiebear — APPROVED** [review]
LGTM

File: options/posix/generic/unistd-stubs.cpp (hunk 3 above)
Comment (Geertiebear) [reply options/posix/generic/unistd-stubs.cpp:22]: There are only two failing conditions for ``setitimer()``, which is if the pointers aren't valid or ``which`` is not one of ``ITIMER_REAL``, ``ITIMER_VIRTUAL``, or ``ITIMER_PROF``, neither of which happens in this case.

**Pull request merged**